demo/popcha-client/web/src/main.py

```diff
@@ -0,0 +1,222 @@
+                    and provider["lao_id"].endswith("="))
+    valid_domain = (("domain" in provider) and ('/' not in provider["domain"])
+                    and not provider["domain"].startswith("http"))
+    valid_public_key = len(provider["public_key"]) > 0
+    return valid_lao_id and valid_domain and valid_public_key
+
+
+def filter_providers(providers: list) -> list:
+    """
+    Keeps only the providers that are valid from the original list
+    :param providers: The original list of providers
+    :return: The valid providers of the original list
+    """
+    return [provider for provider in providers if check_provider(provider)]
+
+
+def load_providers() -> list:
+    """
+    Loads the providers from the configuration file and returns them
+    :return: The providers present in the configuration files
+    """
+    provider_path = path.normpath(
+        path.join(
+            path.dirname(__file__),
+            "../data/providers.json"
+            )
+        )
+    with open(provider_path) as provider_file:
+        return json.loads(provider_file.read())
+
+
+def on_startup() -> None:
+    """
+    Prepare the data needed by the client authentication server (This
+    server). This includes the list of authorized
+    Open ID providers.html, as well as basic config information such as the
+    homepage HTML code or the
+    """
+    global config, core_app, authenticationProvider
+    csrf = CSRFProtect()
+    csrf.init_app(app)
+    core_app = CounterApp()
+    authenticationProvider = Authentication()
+    providers = load_providers()
+    authenticationProvider.providers = filter_providers(providers)
+    config_path = path.normpath(
+        path.join(
+            path.dirname(__file__),
+            "../data/config.json"
+            )
+        )
+    with open(config_path, "r+") as config_file:
+        config = json.loads(config_file.read())
+        if not check_config(config_file):
+            config_file.seek(0)
+            config_file.write(json.dumps(config))
+
+
+on_startup()
+
+
+# Step1: Returns the homepage
+@app.get("/")
+def root() -> str:
+    """
+    Get the homepage.
+    :return: The homepage HTML
+    """
+    providers_string = [f'{provider.get("lao_id")}@{provider.get("domain")}'
+        for provider in authenticationProvider.providers]
+    error: str = ""
+    if "error" in request.args:
+        error = f'Error: {request.args.get("error", default = "", type = str)}'
+    return render_template(
+            'index.html',
+            providers = providers_string,
+            error = error
+            )
+
+
+# Step2: Process user connection data prepare the OIDC request
+@app.route("/authenticate")
+def authentication() -> Response:
+    """
+    Redirect the user to the PoPCHA based authentication server
+    :return: A response which includes a redirect to the original website
+    """
+    provider_id: int = request.args.get("serverAndLaoId", -1, type = int)
```

Comment:
What is the -1 doing here? A default value? Could you also use a named parameter?

Reply:
I will, and yes it is a default value.